### PR TITLE
stabilize dht through data_channel

### PR DESCRIPTION
**Pull request by Raytlty**
use stabilize periodly, find successor and  update finger_table

### CODE REVIEW

File: bns-core/src/types/ice_transport.rs [hunk 1]
```diff
@@ -21,7 +22,7 @@ pub trait IceTransport<Ch: Channel> {
     type ConnectionState;
     type Msg;
 
-    fn new(signaler: Arc<Ch>) -> Self;
+    fn new(signaler: Arc<Ch>, procedures: Arc<Procedures>) -> Self;
```
Comment (RyanKung) [inline bns-core/src/types/ice_transport.rs:25]: Transport is just ONLY for handling connection. Concrete logic should NOT attach to it.

File: bns-core/src/dht/message.rs [hunk 2]
```diff
@@ -0,0 +1,30 @@
+use crate::channels::default::AcChannel;
+use crate::dht::chord::Chord;
+use crate::dht::chord::ChordAction;
+use std::sync::{Arc, Mutex as StdMutex};
+use webrtc::data_channel::RTCDataChannel;
+
+#[derive(Clone, Debug)]
+pub struct Procedures {
+    pub routing: Arc<StdMutex<Chord>>,
```
Comment (RyanKung) [inline bns-core/src/dht/message.rs:9]: `Procedures` or `Successor` is NOT real role, they are virtual and should not include `Chord`, BTW, `Chord` is not routing, it's an algorithm helper and to manage statement for RING.

All function associated to `Procedures` or `Successor` should be implemented under Swarm.

File: bns-core/src/lib.rs [hunk 3]
```diff
@@ -1,6 +1,9 @@
 #![feature(associated_type_defaults)]
 #![feature(async_closure)]
 #![feature(box_syntax)]
+#![allow(dead_code)]
```
Comment (RyanKung) [inline bns-core/src/lib.rs:4]: Pls, Never do this, if a PR should fail CI, let it failed.

File: bns-core/src/types/channel.rs [hunk 4]
```diff
@@ -21,3 +23,17 @@ pub trait Channel {
     async fn send(&self, e: Events) -> Result<()>;
     async fn recv(&self) -> Result<Events>;
 }
+
+// impl Serialize & Deserialize ChordAction to Vec<u8>
+//
+impl Events {
+    pub fn from_action(action: ChordAction) -> Events {
```
Comment (RyanKung) [inline bns-core/src/types/channel.rs:30]: Use `std::From` instead. And it should be a `TryFrom`.

**Review by RyanKung — CHANGES_REQUESTED** [review]
Cant merge yet

File: bns-core/src/dht/message.rs (hunk 2 above)
Comment (Raytlty) [reply bns-core/src/dht/message.rs:9]: `it's an algorithm helper and to manage statement for RING`
why chord cant use for ROUTING

**Comment by Raytlty** [comment]
close this branch, will push another checkout from features/response-msg

**Pull request closed without merging**